src/controllers/user/admin.rs

```diff
@@ -0,0 +1,185 @@
+use axum::{extract::Path, Json};
+use chrono::{NaiveDateTime, Utc};
+use crates_io_database::schema::{emails, users};
+use diesel::{pg::Pg, prelude::*};
+use diesel_async::{scoped_futures::ScopedFutureExt, AsyncConnection, RunQueryDsl};
+use http::request::Parts;
+use utoipa::ToSchema;
+
+use crate::{
+    app::AppState, auth::AuthCheck, models::User, sql::lower, util::errors::AppResult,
+    util::rfc3339, views::EncodableAdminUser,
+};
+
+/// Find user by login, returning the admin's view of the user.
+///
+/// Only site admins can use this endpoint.
+#[utoipa::path(
+    get,
+    path = "/api/v1/users/{user}/admin",
+    params(
+        ("user" = String, Path, description = "Login name of the user"),
+    ),
+    tags = ["admin", "users"],
+    responses((status = 200, description = "Successful Response")),
+)]
+pub async fn get(
+    state: AppState,
+    Path(user_name): Path<String>,
+    req: Parts,
+) -> AppResult<Json<EncodableAdminUser>> {
+    let mut conn = state.db_read_prefer_primary().await?;
+    AuthCheck::only_cookie()
+        .require_admin()
+        .check(&req, &mut conn)
+        .await?;
+
+    get_user(
+        |query| query.filter(lower(users::gh_login).eq(lower(user_name))),
+        &mut conn,
+    )
+    .await
+    .map(Json)
+}
+
+#[derive(Deserialize, ToSchema)]
+pub struct LockRequest {
+    /// The reason for locking the account. This is visible to the user.
+    reason: String,
+
+    /// When to lock the account until. If omitted, the lock will be indefinite.
+    #[serde(default, with = "rfc3339::option")]
+    until: Option<NaiveDateTime>,
+}
+
+/// Lock the given user.
+///
+/// Only site admins can use this endpoint.
+#[utoipa::path(
+    put,
+    path = "/api/v1/users/{user}/lock",
+    params(
+        ("user" = String, Path, description = "Login name of the user"),
+    ),
+    request_body = LockRequest,
+    tags = ["admin", "users"],
+    responses((status = 200, description = "Successful Response")),
+)]
+pub async fn lock(
+    state: AppState,
+    Path(user_name): Path<String>,
+    req: Parts,
+    Json(LockRequest { reason, until }): Json<LockRequest>,
+) -> AppResult<Json<EncodableAdminUser>> {
```

Comment:
for new APIs I would prefer to avoid such "custom action" API endpoints and use a `PATCH` API instead (`PATCH /api/v1/users/{user}/admin`?) like we started for versions too and also to some degree already have for users (using `PUT` unfortunately...).

this will allow us to "easily" extend the functionality of this endpoint in the future instead of having to add new endpoints for basically every field we want to change.